lib/Yield/PTML.cpp

```diff
@@ -246,7 +258,57 @@ tokenTag(std::string &&Buffer, const yield::TagType::Values &Tag) {
   }
 }
 
-static std::string taggedText(const yield::Instruction &Instruction) {
+static TagNode *targetNodeWithSingleImmediateSuccessor(TagNode *Current) {
+  if (Current->Type() != yield::TagType::Target)
+    return nullptr;
+
+  if (Current->successorCount() != 1)
+    return nullptr;
+
+  TagNode *Successor = *Current->successors().begin();
+  if (Successor->Type() != yield::TagType::Immediate)
+    return nullptr;
+
+  if (Current->From() != Successor->From() && Current->To() != Successor->To())
+    return nullptr;
+
+  return Successor;
+}
+
+static std::optional<std::string>
+tryDetectSpecialCases(TagNode *Current,
+                      llvm::StringRef Text,
+                      const yield::BasicBlock &BasicBlock,
+                      const yield::Function &Function,
+                      const model::Binary &Binary) {
```

Comment:
So, here's the situation, first of all, the relation graph (AST thing) does get unrolled bottom up. That is unless we see any "special cases" we want to throw away the internals for.

How I see it, it's a super simple way to make assembly view extensible without changing it too much. I see a "special case" as something akin to "if there is node that satisfies a specific condition, emit it in a special way instead of the normal approach".

For example, here the only special case is a `target` node that contains exactly one successor - `immediate` node and their sizes match (as in, the entire target is a single immediate), if so - we throw away all the information about the immediate and emit a `label` instead.

Other cases I can see us using this in the future is, for example, renaming registers, say for each `register` node on a specific architecture of the binary (i.e. `mips`) emit pretty register names instead of `$N` llvm provides us with. And so on.

That being said, here's a specific example of how we are using this now.
This is what llvm gave us:
```arm
// arm assembly
  b #8
  *-----mnemonic
  --**--immediate
  --**--target
```

If we emit it normally, we end up with
```arm
// arm assembly
  b #8
  *-----mnemonic
  --**--immediate
```
which leads to (simplified)
```html
<span id=mnemonic>b</span> <span id=immediate>#8</span>
```

But, when we run into `target->immediate` structure in our relation graph, we replace the abandoned tree with our own stuff, something like
```arm
// arm assembly
  b my_cool_jump_target_s_name
  *----------------------------mnemonic
  --**************************-label(not a real tag, for illustration purposes only)
```
which ends up emitting
```html
<span id=mnemonic>b</span> <span id=label {all the other cool stuff label has}>my_cool_jump_target_s_name</span>
```

Reply:
1. What you're describing is an AST. You're thinking it's something simpler but it's not, it's an AST, implemented with something more complex than needed (i.e., a graph, which can represent more stuff than what we strictly needed, i.e., loops).
  For the benefit of any future developer and for not inventing our own concepts, if we decide to keep it, we need to call this AST.
2. A well designed AST for our use case should work as follows: 1) build it parsing the input, 2) transform it, 3) emit the output with a single visit without peeking into other nodes.
  If in 3) you need to peek in other nodes, which is what the special case detection thing is doing, you should have done more of 2).
  Basically you have three languages: the input language (LLVM disassembler output), the AST (which is a language on its own) and the output language (PTML). You might think this is a complex read of what's going on but it's what's happening.
4. Now, what described above is not straightforward, and before going this route I really want to be 100% sure we really need the intermediate language (the AST). I've the feeling we can go without any intermediate language since we have full control on the input language (the LLVM disassembler output).
   For your specific example:
   ```
     b #8
   //*-----mnemonic
   //--**--immediate
   //--**--target
   ```
   Can't we simply drop `immediate` annotation and leave `target`? That would solve all our problems, wouldn't it?
   Please post the exact output of the disassembler, unadulterated, with its own tags.

Reply:
1. Sounds good.
2. I guess it sounds good, but I'm still quite a bit hesitant whether "transforming and AST" is a simpler approach than "hey, here's a read-only AST, if you find any nodes you especially like, you can replace them with any string you want in the output". But to each its own I guess. In general, I'm fine with either approach.
3. Yeah, I'd say that's a valid option, but we restrict what we can do with the targets quite a bit of we go this direction. If we imply that "target" tags are only ever emitted for immediates (and call them `address` or something), we limit any chance of ever reusing them in the future, for example, to also mark the `register` targets as targets, you get the point.

> Please post the exact output of the disassembler, unadulterated, with its own tags.

In this case it's just `\tb\t<target:<imm:#8>>`, but that's exactly what we emit, it's not too hard to change it to whatever we want. I'm just hesitant if we REALLY want to limit what we can do with it in the future (including the alternative disassemblers) just for the sake of a dead-simple `LLVM's output`->`our PTML` conversion. Consider that AST would be a perfect decoupling point, as in "to add support for any disassembler you want, you just need to provide conversion to our AST, and we've got it from there".

Reply:
> we heavily misdesigned the input language

Well, this is definitely it. Most of the llvm-side changes are made with "as easy to support / get accepted upstream as possible" mindset. And I don't think we can have as much freedom as I would have wanted coming from that side. Not to mention the eventual point when we start supporting multiple disassemblers.

> What would be the benefit?

Smarter syntax highlighting, better more localized links.

> The current design of `target` is creating more problems than solutions.

Yeah, the more we talk about it the more I hate it. So at least we're on the same page.

> `<imm:target:#8>`

I feel like it's the exact same thing. If this is what we want, we might as well keep the llvm-side the same and then only respect `<target:<imm:#8>>` things and ignore/assert on all the other `target`s. The exact representation is not the problem, I feel like there's a more serious logical implication of trying to shove all the ~~logic~~ magic into a handful of tags.

> It's not dead simple.

By dead-simple here I meant the current approach - graph-less one. What I meant is that if we ever might want to get fancy with the way we reason about it, we should do it as soon as possible, instead of trying to fit all the magic inside a dead-simple converter.

> I think the language of the LLVM backend is using is good enough

:sadparrot:

Reply:
There's an extended discussion on mattermost that eclipses this thread, so I'll close it.